Simplenote/Controllers/PublishController.swift

```diff
@@ -0,0 +1,112 @@
+import Foundation
+
+@objc
+class PublishController: NSObject {
+    private var callbackMap = [String: PublishListenWrapper]()
+    private let timerFactory: TimerFactory
+    private let publishListenerFactory: PublishListenerFactory
+    private var timer: Timer?
+
+    init(timerFactory: TimerFactory = TimerFactory(), publishListenerFactory: PublishListenerFactory = PublishListenerFactory()) {
+        self.timerFactory = timerFactory
+        self.publishListenerFactory = publishListenerFactory
+    }
+
+    func updatePublishState(for note: Note, to published: Bool, completion: @escaping (PublishState) -> Void) {
+        if note.published == published {
+            return
+        }
+
+        callbackMap[note.simperiumKey] = publishListenerFactory.publishListenerWrapper(note: note, block: completion, expiration: Date())
+
+        timer = timerFactory.scheduledTimer(with: Constants.timeOut, completion: {
+            self.removeExpiredCallbacks()
+        })
+
+        changePublishState(for: note, to: published)
+
+        published ?
+            callbackMap[note.simperiumKey]?.update(to: .publishing) : callbackMap[note.simperiumKey]?.update(to: .unpublishing)
```

Comment:
`completion` is usually expected to be invoked **just once**, when the OP is completed (or fails). Otherwise, this API would effectively be coupling two different tasks:

A. Updating the Published State
B. Subscribing to Publishing State changes